xcmds/sshd/sshd.go

```diff
@@ -0,0 +1,86 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"syscall"
+	"unsafe"
+
+	"github.com/gliderlabs/ssh"
+	"github.com/kr/pty" // TODO: get rid of krpty
+	flag "github.com/spf13/pflag"
+)
+
+var (
+	pubKeyFile = flag.StringP("pubkeyfile", "k", "key.pub", "file for public key")
+)
+
+func setWinsize(f *os.File, w, h int) {
+	syscall.Syscall(syscall.SYS_IOCTL, f.Fd(), uintptr(syscall.TIOCSWINSZ),
+		uintptr(unsafe.Pointer(&struct{ h, w, x, y uint16 }{uint16(h), uint16(w), 0, 0})))
+}
+func main() {
+	flag.Parse()
+	publicKeyOption := func(ctx ssh.Context, key ssh.PublicKey) bool {
+		// Glob the users's home directory for all the
+		// possible keys?
+		data, err := ioutil.ReadFile(*pubKeyFile)
+		if err != nil {
+			fmt.Print(err)
+			return false
+		}
+		allowed, _, _, _, _ := ssh.ParseAuthorizedKey(data)
+		return ssh.KeysEqual(key, allowed)
+	}
+
+	server := ssh.Server{
+		LocalPortForwardingCallback: ssh.LocalPortForwardingCallback(func(ctx ssh.Context, dhost string, dport uint32) bool {
+			log.Println("Accepted forward", dhost, dport)
+			return true
+		}),
+		Addr:             ":2222",
+		PublicKeyHandler: publicKeyOption,
+		ReversePortForwardingCallback: ssh.ReversePortForwardingCallback(func(ctx ssh.Context, host string, port uint32) bool {
+			log.Println("attempt to bind", host, port, "granted")
+			return true
+		}),
+		Handler: func(s ssh.Session) {
+			a := s.Command()
+			if len(a) == 0 {
+				a = []string{"bash"}
+			}
+			cmd := exec.Command(a[0], a[1:]...)
```

Comment:
Can this be:

```
cmd := exec.Command("bash", "-c", strings.Join(a, ","))
```

So that you can use pipes and shell features in ssh, for example:

```
ssh example.com 'ls | tac'
```

I think this is how ssh normally behaves.

Reply:
Do we really want to require an intermediate shell for every command we wrong? Doesn't feel right.

Reply:
It's what u-root's other sshd implementation and openssh do, see: https://github.com/u-root/u-root/blob/master/cmds/sshd/sshd.go#L166
It would probably be the least surprising if they all do the same thing.